addons/sourcemod/scripting/nd_discord_playerlist.sp

```diff
@@ -0,0 +1,193 @@
+#include <sdktools>
+#include <nd_stocks>
+#include <nd_team_eng>
+
+Handle SaveTeamsTimer;
+bool SavingTeams;
+Handle DatabaseHandle;
+char PlayerNames[MAXPLAYERS + 1][MAX_NAME_LENGTH];
+
+public Plugin myinfo =
+{
+    name        = "[ND] Discord Player List",
+    author      = "zookatron",
+    description = "Provides player list information to Discord",
+    version     = "dummy",
+    url         = "https://github.com/stickz/Redstone/"
+};
+
+public void OnPluginStart()
+{
+    if (!SQL_CheckConfig("discord"))
+    {
+        SetFailState("Could not find configuration for \"discord\" database");
+        return;
+    }
+
+    SQL_TConnect(OnDatabaseConnected, "discord");
+}
+
+public void ND_OnPlayerTeamChanged(int client, bool valid)
+{
+    SaveTeams();
+}
+
+public void OnClientPutInServer(int client)
+{
+    GetClientName(client, PlayerNames[client], MAX_NAME_LENGTH);
+    SaveTeams();
+}
+
+public void OnClientDisconnect_Post(int client)
+{
+    PlayerNames[client][0] = '\0';
+    SaveTeams();
+}
+
+public OnDatabaseConnected(Handle owner, Handle handle, const char[] error, any:data)
+{
+    if (!IsValidHandle(handle))
+    {
+        SetFailState("Unable to connect to \"discord\" database: %s", error);
+        return;
+    }
+
+    DatabaseHandle = handle;
+
+    char query[1024];
+    FormatEx(query, sizeof(query), "SELECT * FROM players;");
+    SQL_TQuery(DatabaseHandle, Setup_Step1_CheckTable, query);
+}
+
+void Setup_Step1_CheckTable(Handle owner, Handle handle, const char[] error, any:data)
+{
+    // Create the table if it doesn't exist
+    if (error[0] != '\0')
+    {
+        char query[1024];
+        FormatEx(query, sizeof(query), "CREATE TABLE `players` (`id` bigint NOT NULL AUTO_INCREMENT, `name` varchar(128) NOT NULL, `team` varchar(128) NOT NULL, PRIMARY KEY (`id`));");
+        SQL_TQuery(DatabaseHandle, Setup_Step2_CreateTable, query);
+    }
+}
+
+void Setup_Step2_CreateTable(Handle owner, Handle handle, const char[] error, any:data)
+{
+    if (error[0] != '\0')
+    {
+        SetFailState("Unable to create \"players\" table in \"discord\" database: %s", error);
+    }
+}
+
+void SaveTeams()
+{
+    delete SaveTeamsTimer;
+    SaveTeamsTimer = CreateTimer(1.0, Timer_SaveTeams, _, TIMER_FLAG_NO_MAPCHANGE);
+}
+
+public Action Timer_SaveTeams(Handle timer)
+{
+    if(SavingTeams) {
+        SaveTeamsTimer = CreateTimer(1.0, Timer_SaveTeams, _, TIMER_FLAG_NO_MAPCHANGE);
+        return Plugin_Continue;
+    }
+    SaveTeamsTimer = null;
+
+    if(!IsValidHandle(DatabaseHandle)) {
+        LogAction(0, -1, "Database not connected");
+        return Plugin_Continue;
+    }
+
+    SavingTeams = true;
+
+    char query[1024];
+    FormatEx(query, sizeof(query), "TRUNCATE TABLE players;");
+    SQL_TQuery(DatabaseHandle, SaveTeams_Step1_RemovePlayers, query);
+
```

Comment:
How does the Discord side know that the SM plugin is currently re-populating the players table? Wouldn't it be better to use DELETE ... WHERE and INSERT INTO so the players table is always up-to-date?

Reply:
It's literally only a few milliseconds between the truncate and the re-population, I don't think there's any need for increasing the complexity just because the table might be inaccurate for a few milliseconds

Reply:
Makes sense and I suppose if there is some issue it can be dealt with on the Discord side